examples/stabilizer_code.py

```diff
@@ -0,0 +1,318 @@
+            break
+
+        pi = pivot_rows[0]
+        pj = pivot_cols[0]
+
+        # Swap the rows:
+        M[[i, i + pi]] = M[[i + pi, i]]
+
+        # Swap the columns:
+        M[:, [(j + pj), j]] = M[:, [j, (j + pj)]]
+
+        # Since the columns in the left and right half of the matrix represent the same qubit, we
+        # also need to swap the corresponding column in the other half.
+        j_other_half = (j + n) % (2 * n)
+        M[:, [(j_other_half + pj), j_other_half]] = M[:, [j_other_half, (j_other_half + pj)]]
+
+        # Do the elimination.
+        for k in range(i + 1, max_row):
+            if M[k, j] == 1:
+                M[k, :] = np.mod(M[i, :] + M[k, :], 2)
+
+        rank += 1
+
+    # Backward replacing to get identity
+    for r in reversed(range(rank)):
+        i = min_row + r
+        j = min_col + r
+
+        # Do the elimination.
+        for k in reversed(range(min_row, i)):
+            if M[k, j] == 1:
+                M[k, :] = np.mod(M[i, :] + M[k, :], 2)
+
+    return rank
+
+
+class StabilizerCode(object):
+    def __init__(self, group_generators: List[str], allowed_errors: List[str]):
+        n = len(group_generators[0])
+        k = n - len(group_generators)
+
+        # Build the matrix defined in section 3.4. Each row corresponds to one generator of the
+        # code, which is a vector of dimension n. The elements of the vectors are Pauli matrices
+        # encoded as I, X, Y, or Z. However, as described in the thesis, we encode the Pauli
+        # vector of 2*n Booleans.
+        M = np.zeros((n - k, 2 * n), np.int8)
+        for i, group_generator in enumerate(group_generators):
+            for j, c in enumerate(group_generator):
+                if c == 'X' or c == 'Y':
+                    M[i, j] = 1
+                elif c == 'Z' or c == 'Y':
+                    M[i, n + j] = 1
+
+        # Performing the Gaussian elimination as in section 4.1
+        r: int = _GaussianElimination(M, 0, n - k, 0, n)
+        _ = _GaussianElimination(M, r, n - k, n + r, 2 * n)
+
+        # Get matrix sub-components, as per equation 4.3:
+        # A1 = M[0:r, r : (n - k)]
+        A2 = M[0:r, (n - k) : n]
+        # B = M[0:r, n : (n + r)]
+        C1 = M[0:r, (n + r) : (2 * n - k)]
+        C2 = M[0:r, (2 * n - k) : (2 * n)]
+        # D = M[r : (n - k), n : (n + r)]
+        E = M[r : (n - k), (2 * n - k) : (2 * n)]
+
+        X = np.concatenate(
+            [
+                np.zeros((k, r), dtype=np.int8),
+                E.T,
+                np.eye(k, dtype=np.int8),
+                np.mod(E.T @ C1.T + C2.T, 2),
+                np.zeros((k, n - r), np.int8),
+            ],
+            axis=1,
+        )
+
+        Z = np.concatenate(
+            [
+                np.zeros((k, n), dtype=np.int8),
+                A2.T,
+                np.zeros((k, n - k - r), dtype=np.int8),
+                np.eye(k, dtype=np.int8),
+            ],
+            axis=1,
+        )
```

Comment:
I feel it is better to put it into a function like -- `transfer_to_standard_form(M, n, k) -> Tuple[standard_M, X, Z, r] `

Reply:
Done.